compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -307,18 +351,19 @@ class Semantic {
           Result(Hot, Errors.empty)
 
         case Cold =>
-          val error = CallCold(ctor, source, trace)
+          val error = CallCold(ctor, source, trace.toVector)
           Result(Hot, error :: Nil)
 
-        case thisRef: ThisRef =>
-          val value = Warm(klass, outer = thisRef)
-          val res = value.call(ctor, superType = NoType, source)
-          Result(value, res.errors)
-
-        case warm: Warm =>
+        case addr: Addr =>
           // widen the outer to finitize addresses
-          val outer = if warm.outer.isInstanceOf[Warm] then warm.copy(outer = Cold) else warm
+          val outer = addr match
+            case Warm(_, _: Warm) => Cold
+            case _ => addr
+
           val value = Warm(klass, outer)
+          if !heap.contains(value) then
+            val obj = Objekt(klass, fields = mutable.Map.empty, outers = mutable.Map(klass -> outer))
```

Comment:
Will this `Objekt` continue to have empty `fields` and `outers` forever since we never call `init` on the new `value`?

Reply:
The call happens at line 367 -- it's a constructor call on the object.